apiserver/common/instancetypes.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/state"
+)
+
+// InstanceTypeBackend provides the methods necessary to obtain provider
+// information.
+type InstanceTypeBackend interface {
+	environs.EnvironConfigGetter
+	ModelTag() names.ModelTag
+	Close() error
+	GetModel(names.ModelTag) (*state.Model, error)
+}
+
+// modelConfigAble represents a type that can return a *config.Config.
+type modelConfigAble interface {
+	Config() (*config.Config, error)
+}
+
+// newModelEnvironGetterFunc represents a function that can create
+// a modelConfigGetter from an InstanceTypeBackend and any modelConfigAble.
+type newModelEnvironGetterFunc func(InstanceTypeBackend, modelConfigAble) *modelEnvironConfigGetter
+
+// newModelEnvironConfigGetter implements newModelEnvironGetterFunc.
+func newModelEnvironConfigGetter(b InstanceTypeBackend, m modelConfigAble) *modelEnvironConfigGetter {
+	return &modelEnvironConfigGetter{InstanceTypeBackend: b, model: m}
+}
+
+// modelEnvironConfigGetter is a struct that holds a model a backend
+// and can return a modelConfig.
+type modelEnvironConfigGetter struct {
+	InstanceTypeBackend
+	model modelConfigAble
+}
+
+// ModelConfig returns the *config.Config from the modelEnvironConfigGetter
+// model.
+func (st *modelEnvironConfigGetter) ModelConfig() (*config.Config, error) {
+	return st.model.Config()
+}
+
+func toParamsInstanceTypeResult(itypes []instances.InstanceType) []params.InstanceType {
+	result := make([]params.InstanceType, len(itypes))
+	for i, t := range itypes {
+		virtType := ""
+		if t.VirtType != nil {
+			virtType = *t.VirtType
+		}
+		result[i] = params.InstanceType{
+			Name:         t.Name,
+			Arches:       t.Arches,
+			CPUCores:     int(t.CpuCores),
+			Memory:       int(t.Mem),
+			RootDiskSize: int(t.RootDisk),
+			VirtType:     virtType,
+			Deprecated:   t.Deprecated,
+			Cost:         int(t.Cost),
+		}
+	}
+	return result
+}
+
+// NewInstanceTypeConstraints returns an instanceTypeConstraints with the passed
+// parameters.
+func NewInstanceTypeConstraints(st InstanceTypeBackend, constraints constraints.Value, model names.ModelTag) instanceTypeConstraints {
+	return instanceTypeConstraints{
+		backend:     st,
+		constraints: constraints,
+		modelTag:    model,
+
+		environNew:                  environs.New,
+		newModelEnvironConfigGetter: newModelEnvironConfigGetter,
+	}
+}
+
+// instanceTypeConstraints holds necesary params to filter instance types.
+type instanceTypeConstraints struct {
+	backend     InstanceTypeBackend
+	constraints constraints.Value
+	modelTag    names.ModelTag
+
+	newModelEnvironConfigGetter newModelEnvironGetterFunc
+	environNew                  environs.NewEnvironFunc
+}
+
+// InstanceTypes returns a list of the available instance types in the provider according
+// to the passed constraints.
+func InstanceTypes(cons instanceTypeConstraints) (params.InstanceTypesResult, error) {
+	m, err := cons.backend.GetModel(cons.modelTag)
```

Comment:
The fact that we're calling GetModel multiple times for what's always going to be the same model is a bit of a smell I think. As mentioned before, I think we should pass the CloudSpec (perhaps Environ?) in.

If this were split out, then it would be reasonable to fail the entire API call if we fail to get the CloudSpec in the case of the MachineManager facade. For the Cloud facade, it would be reasonable to fail the entire API call if we fail to get the Model; and then fail individual results if we can't get the CloudSpec (e.g. if an invalid CloudTag or region is specified).